File: cdk/lambda/vectorIndexManagerSigV4/main.js
const { Client } = require("@opensearch-project/opensearch");
const { AwsSigv4Signer } = require("@opensearch-project/opensearch/aws");
const { defaultProvider } = require("@aws-sdk/credential-provider-node");

const MAX_FORBIDDEN_RETRY_WINDOW_MS = 5 * 60 * 1000;
const FORBIDDEN_RETRY_SLEEP_MS = 15 * 1000;
const INDEX_STABILIZE_MAX_WINDOW_MS = 3 * 60 * 1000;
const INDEX_STABILIZE_POLL_MS = 10 * 1000;
const INDEX_STABILIZE_EXTRA_DELAY_MS = 20 * 1000;

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

function normalizeEndpoint(endpoint) {
  const trimmed = String(endpoint || "").trim();
  if (!trimmed) {
    throw new Error("CollectionEndpoint is required");
  }
  const withProtocol = /^https?:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`;
  return withProtocol.replace(/\/$/, "");
}

function buildClient(endpoint, region) {
  const credentialProvider = defaultProvider();

  return new Client({
    ...AwsSigv4Signer({
      region,
      service: "aoss",
      getCredentials: () => credentialProvider(),
    }),
    node: endpoint,
    maxRetries: 0,
    requestTimeout: 30000,
  });
}

function extractStatusCode(error) {
  return error?.statusCode || error?.meta?.statusCode || 0;
}

function extractErrorText(error) {
  const body = error?.meta?.body;
  if (typeof body === "string") {
    return body;
  }
  if (body) {
    return JSON.stringify(body);
  }
  return String(error?.message || "");
}

function responseBodyValue(response) {
  if (typeof response === "boolean") {
    return response;
  }
  if (response && Object.prototype.hasOwnProperty.call(response, "body")) {
    return response.body;
  }
  return response;
}

async function indexExists(client, indexName) {
  const response = await client.indices.exists({ index: indexName });
  const value = responseBodyValue(response);

  if (typeof value === "boolean") {
    return value;
  }

  if (typeof response?.statusCode === "number") {
    return response.statusCode === 200;
  }

  return Boolean(value);
}

async function ensureIndexWithPropagation(client, props) {
  const {
    indexName,
    vectorField,
    textField,
    metadataField,
    dimensions,
  } = props;

  const deadline = Date.now() + MAX_FORBIDDEN_RETRY_WINDOW_MS;
  let attempt = 0;

  const body = {
    settings: {
      index: {
        knn: true,
      },
    },
    mappings: {
      properties: {
        [vectorField]: {
          type: "knn_vector",
          dimension: dimensions,
          method: {
            engine: "faiss",
            name: "hnsw",
            space_type: "l2",
          },
        },
        [textField]: {
          type: "text",
        },
        [metadataField]: {
          type: "text",
          index: false,
        },
      },
    },
  };

  while (Date.now() < deadline) {
    attempt += 1;

    try {
      if (await indexExists(client, indexName)) {
        console.log(`Index '${indexName}' already exists. Skipping create.`);
        return;
      }

      await client.indices.create({
        index: indexName,
        body,
      });

      console.log(`Index '${indexName}' created successfully on attempt ${attempt}.`);
      return;
    } catch (error) {
      const statusCode = extractStatusCode(error);
      const errorText = extractErrorText(error);
      const forbidden = statusCode === 403 || /forbidden/i.test(errorText);
      const alreadyExists = /resource_already_exists_exception|already exists/i.test(errorText);

      if (alreadyExists) {
        console.log(`Index '${indexName}' already exists (race condition). Treating as success.`);
        return;
      }

      if (forbidden) {
        const secondsLeft = Math.max(0, Math.floor((deadline - Date.now()) / 1000));
        console.log(
          `Attempt ${attempt} received Forbidden (403/ResponseError). ` +
          `Waiting 15s for AOSS policy propagation. Remaining retry window: ${secondsLeft}s.`
        );
        await sleep(FORBIDDEN_RETRY_SLEEP_MS);
        continue;
      }

      throw error;
    }
  }

  throw new Error(
    `Timed out after 5 minutes waiting for AOSS policy propagation while creating index '${indexName}'.`
  );
}

async function waitForIndexStabilization(client, indexName) {
  const deadline = Date.now() + INDEX_STABILIZE_MAX_WINDOW_MS;
  let attempt = 0;

  while (Date.now() < deadline) {
    attempt += 1;
    try {
      const exists = await indexExists(client, indexName);
      if (!exists) {
        console.log(`Index stabilization attempt ${attempt}: index not visible yet.`);
        await sleep(INDEX_STABILIZE_POLL_MS);
        continue;
      }

      await client.indices.get({ index: indexName });
      console.log(`Index '${indexName}' is visible and retrievable. Applying final delay for Bedrock consistency.`);
      await sleep(INDEX_STABILIZE_EXTRA_DELAY_MS);
      return;
    } catch (error) {
      const statusCode = extractStatusCode(error);
      const errorText = extractErrorText(error);
      const transient = statusCode === 404 || statusCode === 403 || /no such index|forbidden/i.test(errorText);

      if (transient) {
        console.log(`Index stabilization attempt ${attempt}: status=${statusCode}, waiting for AOSS data-plane consistency.`);
        await sleep(INDEX_STABILIZE_POLL_MS);
        continue;
      }

      throw error;
    }
  }

  throw new Error(`Timed out waiting for index '${indexName}' to become visible to downstream services.`);
}

exports.handler = async (event) => {
  console.log("VectorIndexManager event:", JSON.stringify(event));

  const requestType = event?.RequestType;
  const existingPhysicalId = event?.PhysicalResourceId || "vector-index-custom-resource";

  if (requestType === "Delete") {
    return {
      PhysicalResourceId: existingPhysicalId,
      Data: {
        SkippedDelete: "true",
      },
    };
  }

  const props = event?.ResourceProperties || {};
  const endpoint = normalizeEndpoint(props.CollectionEndpoint);
  const region = props.Region;
  const indexName = props.IndexName;

  if (!region || !indexName) {
    throw new Error("Region and IndexName are required resource properties");
  }

  const client = buildClient(endpoint, region);

  await ensureIndexWithPropagation(client, {
    indexName,
    vectorField: props.VectorField,
    textField: props.TextField,
    metadataField: props.MetadataField,
    dimensions: Number(props.Dimensions),
  });

  await waitForIndexStabilization(client, indexName);

  return {
    PhysicalResourceId: existingPhysicalId === "vector-index-custom-resource"
      ? `${indexName}-vector-index`
      : existingPhysicalId,
    Data: {
      IndexName: indexName,
    },
  };
};

File: cdk/lambda/vectorIndexManagerSigV4/main.py
import json
import logging
import time
from urllib.parse import urlparse

import boto3
from opensearchpy import AWSV4SignerAuth, OpenSearch, RequestsHttpConnection, exceptions

logger = logging.getLogger()
logger.setLevel(logging.INFO)

MAX_FORBIDDEN_RETRY_WINDOW_SECONDS = 5 * 60
FORBIDDEN_RETRY_SLEEP_SECONDS = 15
INDEX_STABILIZE_MAX_WINDOW_SECONDS = 3 * 60
INDEX_STABILIZE_POLL_SECONDS = 10
INDEX_STABILIZE_EXTRA_DELAY_SECONDS = 20


def sleep(seconds: int) -> None:
    time.sleep(seconds)


def normalize_endpoint(endpoint: str) -> str:
    trimmed = str(endpoint or "").strip()
    if not trimmed:
        raise ValueError("CollectionEndpoint is required")
    if not trimmed.startswith("https://") and not trimmed.startswith("http://"):
        trimmed = f"https://{trimmed}"
    return trimmed.rstrip("/")


def build_client(endpoint: str, region: str) -> OpenSearch:
    credentials = boto3.Session().get_credentials()
    auth = AWSV4SignerAuth(credentials, region, "aoss")

    parsed = urlparse(endpoint)
    host = parsed.netloc or parsed.path

    return OpenSearch(
        hosts=[{"host": host, "port": 443}],
        http_auth=auth,
        use_ssl=True,
        verify_certs=True,
        connection_class=RequestsHttpConnection,
        timeout=30,
        retry_on_timeout=False,
        max_retries=0,
    )


def error_message(error: Exception) -> str:
    info = getattr(error, "info", None)
    if isinstance(info, str):
        return info
    if info is not None:
        try:
            return json.dumps(info)
        except Exception:
            return str(info)
    return str(error)


def is_index_already_exists(error_text: str) -> bool:
    text = error_text.lower()
    return "resource_already_exists_exception" in text or "already exists" in text


def ensure_index_with_propagation(client: OpenSearch, props: dict) -> None:
    index_name = props["indexName"]
    vector_field = props["vectorField"]
    text_field = props["textField"]
    metadata_field = props["metadataField"]
    dimensions = int(props["dimensions"])

    deadline = time.time() + MAX_FORBIDDEN_RETRY_WINDOW_SECONDS
    attempt = 0

    body = {
        "settings": {
            "index": {
                "knn": True
            }
        },
        "mappings": {
            "properties": {
                vector_field: {
                    "type": "knn_vector",
                    "dimension": dimensions,
                    "method": {
                        "engine": "faiss",
                        "name": "hnsw",
                        "space_type": "l2"
                    }
                },
                text_field: {
                    "type": "text"
                },
                metadata_field: {
                    "type": "text",
                    "index": False
                }
            }
        }
    }

    while time.time() < deadline:
        attempt += 1

        try:
            if client.indices.exists(index=index_name):
                logger.info("Index '%s' already exists. Skipping create.", index_name)
                return

            client.indices.create(index=index_name, body=body)
            logger.info("Index '%s' created successfully on attempt %s.", index_name, attempt)
            return
        except exceptions.TransportError as error:
            status_code = getattr(error, "status_code", 0)
            text = error_message(error)
            forbidden = status_code == 403 or "forbidden" in text.lower()

            if is_index_already_exists(text):
                logger.info("Index '%s' already exists (race condition). Treating as success.", index_name)
                return

            if forbidden:
                seconds_left = max(0, int(deadline - time.time()))
                logger.warning(
                    "Attempt %s received Forbidden (403/TransportError). Waiting %ss for AOSS policy propagation. Remaining retry window: %ss.",
                    attempt,
                    FORBIDDEN_RETRY_SLEEP_SECONDS,
                    seconds_left,
                )
                sleep(FORBIDDEN_RETRY_SLEEP_SECONDS)
                continue

            raise

    raise RuntimeError(
        f"Timed out after 5 minutes waiting for AOSS policy propagation while creating index '{index_name}'."
    )


def wait_for_index_stabilization(client: OpenSearch, index_name: str) -> None:
    deadline = time.time() + INDEX_STABILIZE_MAX_WINDOW_SECONDS
    attempt = 0

    while time.time() < deadline:
        attempt += 1
        try:
            exists = client.indices.exists(index=index_name)
            if not exists:
                logger.info("Index stabilization attempt %s: index not visible yet.", attempt)
                sleep(INDEX_STABILIZE_POLL_SECONDS)
                continue

            client.indices.get(index=index_name)
            logger.info(
                "Index '%s' is visible and retrievable. Applying final delay for Bedrock consistency.",
                index_name,
            )
            sleep(INDEX_STABILIZE_EXTRA_DELAY_SECONDS)
            return
        except exceptions.TransportError as error:
            status_code = getattr(error, "status_code", 0)
            text = error_message(error).lower()
            transient = status_code in (403, 404) or "no such index" in text or "forbidden" in text

            if transient:
                logger.info(
                    "Index stabilization attempt %s: status=%s, waiting for AOSS data-plane consistency.",
                    attempt,
                    status_code,
                )
                sleep(INDEX_STABILIZE_POLL_SECONDS)
                continue

            raise

    raise RuntimeError(f"Timed out waiting for index '{index_name}' to become visible to downstream services.")


def handler(event, context):
    logger.info("VectorIndexManager event: %s", json.dumps(event))

    request_type = event.get("RequestType")
    existing_physical_id = event.get("PhysicalResourceId") or "vector-index-custom-resource"

    if request_type == "Delete":
        return {
            "PhysicalResourceId": existing_physical_id,
            "Data": {
                "SkippedDelete": "true"
            }
        }

    props = event.get("ResourceProperties") or {}
    endpoint = normalize_endpoint(props.get("CollectionEndpoint"))
    region = props.get("Region")
    index_name = props.get("IndexName")

    if not region or not index_name:
        raise ValueError("Region and IndexName are required resource properties")

    client = build_client(endpoint, region)

    ensure_index_with_propagation(
        client,
        {
            "indexName": index_name,
            "vectorField": props.get("VectorField"),
            "textField": props.get("TextField"),
            "metadataField": props.get("MetadataField"),
            "dimensions": int(props.get("Dimensions")),
        },
    )

    wait_for_index_stabilization(client, index_name)

    physical_resource_id = existing_physical_id
    if existing_physical_id == "vector-index-custom-resource":
        physical_resource_id = f"{index_name}-vector-index"

    return {
        "PhysicalResourceId": physical_resource_id,
        "Data": {
            "IndexName": index_name
        }
    }
